### PR TITLE
Add cache to target client

### PR DESCRIPTION
**Pull request by tim-ebert**
**What this PR does / why we need it**:
This PR adds a cache to the client for the target cluster. That way, `grm`'s health controller will not retrieve all objects of the manged resource from the API server (by default every minute), but rather use the cache instead. Additionally it only retrieves the resource, if it is of a `Kind`, that will actually be checked for its health.

Also, `grm` now does not set `.status.resources[].namespace` to `default` for cluster scoped resources (e.g. `CustomResourceDefinitions`). The previous behaviour seemed to cause problems, when retrieving such resources from the cached client, so I changed it along the way.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/1953

I did some rudimentary before/after comparison via prometheus in my local setup with the following results:
- Received bytes for
  - `grm` in `garden` namespace down by `~65%` (`~15.5kB/s` -> `~5.5kB/s`)
  - `grm` in each shoot namespace down by `~30%` (`~16.9kB/s` -> `~11.8kB/s`) (should not really contribute much to external traffic, as most of the managed resources here target the Shoot's API server (in-cluster))

- Sent bytes for
  - `grm` in `garden` namespace down by `~41%` (`~2.81kB/s` -> `~1.65kB/s`)
  - `grm` in each shoot namespace down by `~14%` (`~7.31kB/s` -> `~6.31kB/s`) (again, should not really contribute much to external traffic, as most of the managed resources here target the Shoot's API server (in-cluster))

**Special notes for your reviewer**:

**Release note**:

```improvement operator
`gardener-resource-manager` now makes use of a caching client for talking to the targeted API server, which reduces its network traffic.
```

### CODE REVIEW

File: pkg/controller/managedresources/health/health_checker.go [hunk 1]
```diff
@@ -15,64 +15,85 @@
...
+	appsv1.SchemeGroupVersion.WithKind("DaemonSet").GroupKind(),
+	appsv1.SchemeGroupVersion.WithKind("Deployment").GroupKind(),
+	batchv1.SchemeGroupVersion.WithKind("Job").GroupKind(),
+	corev1.SchemeGroupVersion.WithKind("Pod").GroupKind(),
+	appsv1.SchemeGroupVersion.WithKind("ReplicaSet").GroupKind(),
+	corev1.SchemeGroupVersion.WithKind("ReplicationController").GroupKind(),
+	appsv1.SchemeGroupVersion.WithKind("StatefulSet").GroupKind(),
+}
+
+func CanBeChecked(groupKind schema.GroupKind) bool {
```
Comment (rfranzke) [inline pkg/controller/managedresources/health/health_checker.go]: Doc string for exported function

File: pkg/controller/managedresources/health/health_controller.go [hunk 2]
```diff
@@ -72,18 +72,35 @@ func (r *HealthReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 
 	resourcesObjectReferences := mr.Status.Resources
 	for _, ref := range resourcesObjectReferences {
-		obj := &unstructured.Unstructured{}
-		obj.SetAPIVersion(ref.APIVersion)
-		obj.SetKind(ref.Kind)
+		// only retrieve object if we will be able to check its health
```
Comment (rfranzke) [inline pkg/controller/managedresources/health/health_controller.go]: Hm, can't we at least check if the object exists instead of silently continuing?

Comment (tim-ebert) [reply pkg/controller/managedresources/health/health_controller.go]: Good catch, fixed, PTAL :)

**Comment by tim-ebert** [comment]
/hold
I have one more optimisation I want to make.

**Comment by tim-ebert** [comment]
/hold cancel
The optimisation I wanted to make isn't really relevant.

File: cmd/gardener-resource-manager/app/app.go [hunk 3]
```diff
@@ -206,17 +236,33 @@ func NewControllerManagerCommand(ctx context.Context) *cobra.Command {
...
+			go func() {
+				if err := targetCache.Start(ctx.Done()); err != nil {
+					entryLog.Error(err, "error syncing target cache")
+					os.Exit(1)
+				}
+			}()
+
+			targetCache.WaitForCacheSync(ctx.Done())
+
+			go func() {
```
Comment (timuthy) [inline cmd/gardener-resource-manager/app/app.go:249]: I wonder if we should synchronize this routine with the main one in order to wait for the manager to be finished. WDYT?

Comment (tim-ebert) [reply cmd/gardener-resource-manager/app/app.go:249]: Can you elaborate, what you mean by "synchronize this routine with the main one"? 
The manager actually does exactly the same as our main routine, it waits for the `ctx.Done()` channel to be closed (or an error happened) and stops afterwards.

Comment (tim-ebert) [reply cmd/gardener-resource-manager/app/app.go:249]: But we could also run `mgr.Start()` directly in the main routine (as it was before) as it blocks until the stop channel is closed, if that is what you mean.

**Comment by tim-ebert** [comment]
@timuthy and I reworked `app.go` to not use `os.Exit()` and return an error instead, and to use a new context and properly stop the manager or the cache when either errors.
PTAL :)

**Comment by tim-ebert** [comment]
Rebased to resolve conflicts.
@rfranzke do you also want to take another look, or can we get this in?

**Pull request merged**